### PR TITLE
Migrate legacy fp32 matmul precision readers to per-backend API

### PR DESCRIPTION
**Pull request by jeffdaily**
Reading `torch.get_float32_matmul_precision()` or `torch.backends.cuda.matmul.allow_tf32` raises a mixed-API TORCH_CHECK error whenever fp32 precision was last written through the new per-backend `torch.backends.cuda.matmul.fp32_precision` API, because the new-style setter intentionally does not update the legacy global enum. Several framework-internal hot paths still read through the legacy getters, so a single new-style write poisons inductor compilation (cache keys in codecache and select_algorithm), ROCm CK instance gating, flex attention precision selection, and the benchmark compile util. This blocks migrating tests and user code to the new API, as suggested in review on #185236.

This PR converts those framework-internal readers to the resolved per-backend getter (`torch._C._get_fp32_precision_getter`), which never throws and is the same source of truth the aten matmul paths already consult. The converted readers are the two inductor cache keys (now keyed on the (cuda, mkldnn) matmul precision pair via a shared `fp32_matmul_precision_key()` helper; the pair carries exactly the information of the legacy string: "highest" -> ieee/ieee, "high" -> tf32/tf32, "medium" -> tf32/bf16), ROCm CK instance gating, flex attention's `set_float32_precision()` (its legacy fallback for the unset "none" state still threw on new-style writes to other backends; a resolved "none" means unset at every level of the per-backend hierarchy, which is exactly the legacy default "highest", so the fallback collapses to "ieee"), and `torch/utils/benchmark/utils/compile.py`, which previously called the legacy getter at import time and now snapshots lazily. The legacy API surface itself keeps its consistency check by design, and legacy writers are untouched since the legacy setters keep both representations in sync. A regression test (`test_fp32_precision_new_api_internal_readers` in test_cuda.py) exercises the converted readers under the mixed state that makes the legacy getters raise.

The TunableOp TypeName and dynamo `get_global_state_key()` readers this PR originally also migrated were converted upstream in the meantime by #195301, so they dropped out of the diff when rebasing.

#196860 landed a narrower workaround at the two autotune cache-key sites this PR migrates: it reads the new-style per-backend value and only falls back to the legacy getter when that value is not `bfx9`. That dodges the throw for `bfx9` specifically, but the legacy getter still throws for every other precision written through the new API. `fp32_matmul_precision_key()` supersedes it, because it never consults the legacy getter at all and so covers `bfx9` along with the rest; merging main therefore drops that workaround. `test_precompile_cache_key_handles_bfx9` is updated for the pair-format key and tightened to assert the legacy getter is never called for any precision.

Compatibility note: the autotune/precompile cache-key component changes format (e.g. "high" becomes "cuda:tf32,mkldnn:tf32"), which invalidates existing on-disk autotune caches once on upgrade; this causes a one-time re-benchmark and has no correctness impact.

An alternative was to relax the TORCH_CHECK in `Context::allowTF32CuBLAS()` / `Context::float32MatmulPrecision()`, but that check is the only guard alerting users to genuinely inconsistent mixed usage, so migrating the internal readers is strictly safer.

A follow-up will migrate the module-level precision setup in the test suite to the per-backend API (the follow-up promised in the #185236 review thread).

Authored with an AI assistant (Claude Code).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jeffdaily** [comment]
Both review points addressed in eda5fdab52d:

- `get_float32_precision()` in flex_attention.py now uses the per-backend getter directly. I verified the flagged scenario empirically: a mkldnn-only new-style write (`torch.backends.mkldnn.matmul.fp32_precision = "bf16"`) throws in the old fallback and works after the fix. Dropping the legacy fallback also loses nothing for the never-set case: `TORCH_ALLOW_TF32_CUBLAS_OVERRIDE` initializes the per-backend map itself (see the `fp32_precision` initializer in Context.h), so the resolved getter reflects it.
- Added `test_fp32_precision_new_api_internal_readers` to test_cuda.py next to `test_invalid_status_for_legacy_api`: it puts the process into the exact mixed state that the legacy getters raise on, then asserts the converted readers (`get_global_state_key`, `fp32_matmul_precision_key`, flex `get_float32_precision`) work and return the new-style values. Confirmed the test scenario raises on an unfixed build at `get_global_state_key()`.
- PR description updated with the cache-key compatibility note.

The TunableGemm.h path is not unit-tested here since it needs TunableOp on MI300 hardware; the existing `test_tf32_tunableop` exercises it and will pin the new-API behavior once the follow-up test-suite migration switches those tests to per-backend writes.

**Comment by jeffdaily** [comment]
@albanD @jansel the PR failures here are caused by a race introduced by the fp32/tf32 state leak detector PR https://github.com/pytorch/pytorch/pull/185236.  Rather than revert, I have a forward fix PR https://github.com/pytorch/pytorch/pull/190285 here if you could take a look.

EDIT: I was operating on stale information.  #185236 had been auto-reverted.  So I've rolled #190285  into it to fix the reason for the auto-revert.  Will rebase this PR once it relands.

**Comment by drisspg** [comment]
https://github.com/pytorch/pytorch/blob/da9c79b096d4b842c9fe4ac42298380a78cf69f3/torch/_inductor/kernel/flex/flex_attention.py#L130-L142

does this work correctly with ^^ ?

**Comment by jeffdaily** [comment]
Yes. `_get_fp32_precision_getter("cuda", "matmul")` returns the resolved per-backend value, which matches the linked logic in every state: `highest` -> `ieee`, `high`/`medium` -> `tf32`, and the default unset (`none`) state -> `ieee` (the old code reached the same result via the `get_float32_matmul_precision() == "highest"` fallback). The difference is that the old fallback throws the mixed-API `TORCH_CHECK` when precision was set through the new per-backend API on another backend (e.g. `torch.backends.mkldnn.matmul.fp32_precision = "bf16"` leaves cuda/matmul at `none` and makes the legacy enum inconsistent); the per-backend getter returns the correct value there instead of raising. I verified all of these empirically.

_This reply was drafted with an AI assistant (Claude Code)._

File: torch/utils/benchmark/utils/compile.py [hunk 1]
```diff
@@ -26,18 +26,24 @@
...
 if HAS_TABULATE:
     def _enable_tensor_cores() -> None:
-        global _warned_tensor_cores
+        global _warned_tensor_cores, _default_float_32_precision
 
         if torch.cuda.is_available():
-            if torch.backends.cuda.matmul.allow_tf32 is False and torch.cuda.get_device_capability() >= (8, 0):
+            tf32_off = torch._C._get_fp32_precision_getter("cuda", "matmul") != "tf32"
+            if tf32_off and torch.cuda.get_device_capability() >= (8, 0):
+                # tf32 being off implies the legacy precision was "highest"
```
Comment (drisspg) [inline torch/utils/benchmark/utils/compile.py:34]: > **AI-assisted review note:** Could we snapshot and restore the per-backend precision values directly here? The resolved CUDA state being non-TF32 does not imply the legacy state was `highest`. For example, starting with CUDA `ieee` and MKLDNN `bf16`, this sets both backends to `tf32`, then `_disable_tensor_cores()` restores both to `ieee`, silently losing the caller's MKLDNN setting.

File: torch/_inductor/utils.py [hunk 2]
```diff
@@ -2977,6 +2977,14 @@ def parallel_num_threads() -> int:
     return threads
 
 
+def fp32_matmul_precision_key() -> str:
+    # Read per-backend fp32 precision instead of
+    # torch.get_float32_matmul_precision(), which raises if the legacy and
+    # per-backend APIs have been mixed.
+    getter = torch._C._get_fp32_precision_getter
+    return f"cuda:{getter('cuda', 'matmul')},mkldnn:{getter('mkldnn', 'matmul')}"
+
```
Comment (drisspg) [inline torch/_inductor/utils.py:2996]: > **AI-assisted review note:** Should this migration also update the nearby XPU path in `get_device_tflops()`? It still reads `torch.backends.mkldnn.allow_tf32`, so a new-style `torch.backends.mkldnn.matmul.fp32_precision = "tf32"` is not reflected in the roofline/device estimate. Using `_get_fp32_precision_getter("mkldnn", "matmul") == "tf32"` there would keep this path consistent with the new API.

**Comment by jeffdaily** [comment]
Friendly ping on this one. @drisspg I addressed your question about `get_float32_precision` above (the per-backend getter is behavior-equivalent to the old logic in every state and additionally avoids the mixed-API `TORCH_CHECK` throw). CI was green on the prior head before I just refreshed the branch onto viable/strict (rerunning now); the diff is scoped to the framework-internal readers of the legacy matmul-precision API. @albanD this is the reader-migration prerequisite for the test-suite `fp32_precision` migration you suggested on #185236 -- an approval from either of you would let it land and unblock that follow-up.

_This message was drafted with an AI assistant (Claude Code)._

**Review by drisspg — APPROVED** [review]

File: torch/_inductor/kernel/flex/flex_attention.py [hunk 3]
```diff
@@ -129,11 +129,7 @@ def flex_attention_grid(batch_size, q_heads, num_queries, d_model, meta, *, cdiv
 
 def get_float32_precision():
     if (
-        (
-            torch.backends.cuda.matmul.fp32_precision == "ieee"
-            if torch.backends.cuda.matmul.fp32_precision != "none"
-            else torch.get_float32_matmul_precision() == "highest"
-        )
```
Comment (lakshayg) [inline torch/_inductor/kernel/flex/flex_attention.py:136]: Wow! This was one hard to read conditional ☠️

If it's not too hard to do, it would be nice to have a unit test for this function.

File: test/test_cuda.py [hunk 4]
```diff
@@ -1334,6 +1334,23 @@ def test_invalid_status_for_legacy_api(self):
             with self.assertRaisesRegex(RuntimeError, "mix of the legacy and new APIs"):
                 print(torch.backends.cuda.matmul.allow_tf32)
 
+    @recover_orig_fp32_precision
+    @serialTest()
+    def test_fp32_precision_new_api_internal_readers(self):
+        # Framework-internal readers must use the per-backend getters and stay
+        # usable after new-style writes put the legacy enum out of sync (the
```
Comment (lakshayg) [inline test/test_cuda.py:1342]: Since it's possible for legacy enums to go out of sync, should we just get rid of them altogether? Doesn't need to be this PR but I would like to understand if there's any good reason to keep them around. If we decide they can be removed then marking them `[[deprecated]]` would be a good first move.

**Comment by janbernloehr** [comment]
Adding concrete downstream impact: this fixes the failure reported in #196728, and the fix has been validated against the original workload.

The reproduction was a DeepSeek-V3 16B pretraining workload with:

- PyTorch `2.15.0a0+gita56a3a18`
- TorchTitan `0.2.2+gitc215a12`
- Triton `3.8.0`
- CUDA 13.x on aarch64
- 8x GB200 across 2 nodes (4 GPUs per node)

During distributed initialization, TorchTitan sets

```python
torch.backends.cuda.matmul.fp32_precision = "bfx9"
```

on compute capability 10.0 and newer. The failure occurs even without explicitly enabling model compilation because TorchTitan's determinism setup compiles FlexAttention. The first compiled FlexAttention call reaches `PersistentCache.lookup`, calls the legacy `torch.get_float32_matmul_precision()`, and raises the mixed legacy/new API error before training step 1—even though the FlexAttention inputs are bf16 and the user never called the legacy precision API.

Applying the on-CUDA-path hunks from this PR to the PyTorch installation used by the original workload resolved the failure. All three previously failing runs then trained for 50 steps to completion:

- Run 1: average loss 6.696
- Run 2: average loss 6.691
- Run 3: average loss 6.736

Across the three runs there were 0 OOMs and 0 allocation retries.

The changes here to `PersistentCache.lookup` and `create_precompile_key` address exactly that failing path. The resolved `(cuda, mkldnn)` key is also preferable to the narrower helper proposed in #196737: it preserves the distinction between legacy `high` and `medium`, and it avoids falling back to the throwing legacy getter when CUDA is unset but another backend was configured through the new API.

Maybe it is useful to add `bfx9` coverage to the regression test, incorporating the motivating case from #196728/#196737.

**Comment by drisspg** [comment]
@jeffdaily any blockers here?

**Comment by jeffdaily** [comment]
The blocker was a conflict with #196860, which landed a narrower fix at the same two autotune cache-key sites this PR touches: it reads the new-style per-backend value and only falls back to the legacy getter when that value is not `bfx9`. That avoids the throw for `bfx9`, but the legacy getter still throws for any other precision written through the new API.

I resolved it by taking `fp32_matmul_precision_key()` at both sites and dropping that workaround, since the helper never consults the legacy getter at all and so covers `bfx9` along with every other mixed-API state. `test_precompile_cache_key_handles_bfx9` is updated for the pair-format key and tightened to assert the legacy getter is never called for any precision, rather than only for `bfx9`.

@eddieyan flagging you since this removes the workaround from #196860.

Merge is pushed and CI is rerunning. Your approval still stands from 2026-08-12; let me know if you want another look at the conflict resolution given it deletes code from #196860.

_This reply was drafted with an AI assistant (Claude Code)._

**Comment by jeffdaily** [comment]
@pytorchbot merge -i

The three red checks are all unrelated. Two are `inductor_aoti_fallback` shards marked `unstable`. The third, `test_nested_region_inductor_config_max_autotune_parent_max_autotune_True_nested_max_autotune_False`, fails identically on trunk: the two most recent `rocm-mi300` runs on main both fail that test plus the `True_True` parametrization, while both `parent=False` cases pass. Dr.CI flagged it unclassified only because the workflow did not run on this PR's merge base.

_This comment was drafted with an AI assistant (Claude Code)._
